// jwt-redis/build/JWTRedis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsonwebtoken = require("jsonwebtoken");
const Redis_1 = require("./Redis");
const TokenInvalidError_1 = require("./error/TokenInvalidError");
const TokenDestroyedError_1 = require("./error/TokenDestroyedError");
class JWTRedis {
    constructor(redisClient, options) {
        this.redisClient = redisClient;
        this.options = Object.assign({ prefix: 'jwt_label:' }, options || {});
        this.redis = new Redis_1.default(redisClient);
    }
    sign(payload, secretOrPrivateKey, options) {
        return Promise.resolve()
            .then(async () => {
            const jti = payload.jti || generateId(10);
            const token = jsonwebtoken.sign(Object.assign({}, payload, { jti }), secretOrPrivateKey, options);
            const decoded = jsonwebtoken.decode(token);
            const key = this.options.prefix + jti;
            if (decoded.exp) {
                await this.redis.setExp(key, 'true', 'EX', Math.floor(decoded.exp - Date.now() / 1000));
            }
            else {
                await this.redis.set(key, 'true');
            }
            return token;
        });
    }
    destroy(jti) {
        const key = this.options.prefix + jti;
        return this.redis.del(key);
    }
    decode(token, options) {
        return jsonwebtoken.decode(token, options);
    }
    verify(token, secretOrPublicKey, options) {
        return new Promise((resolve, reject) => {
            return jsonwebtoken.verify(token, secretOrPublicKey, options, (err, decoded) => {
                if (err) {
                    return reject(err);
                }
                return resolve(decoded);
            });
        }).then((decoded) => {
            if (!decoded.jti) {
                throw new TokenInvalidError_1.default();
            }
            const key = this.options.prefix + decoded.jti;
            return this.redis.get(key)
                .then((result) => {
                if (!result) {
                    throw new TokenDestroyedError_1.default();
                }
                return decoded;
            });
        });
    }
}
exports.default = JWTRedis;
function generateId(length) {
    let result = '';
    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    const charactersLength = characters.length;
    for (var i = 0; i < length; i++) {
        result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
}
